### PR TITLE
[RFC] add journal bindings around sd-journal

### PR DESCRIPTION
**Pull request by jonboulle**
Work in progress. Based on Dan Mace's initial prototype announced here:
http://lists.freedesktop.org/archives/systemd-devel/2014-June/019840.html

Throwing up to start some discussion...

### CODE REVIEW

File: journal/journal.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func (j *Journal) GetData(field string) (string, error) {
+	f := C.CString(field)
+	defer C.free(unsafe.Pointer(f))
+
+	var d unsafe.Pointer
+	var l C.size_t
+
+	err := C.sd_journal_get_data(j.cjournal, f, &d, &l)
```
Comment (vcaputo) [inline journal/journal.go]: From the sd_journal_get_data() documentation:
> The returned data is in a read-only memory map and is only valid until the next invocation of sd_journal_get_data() or sd_journal_enumerate_data(), or the read pointer is altered.

Doesn't this imply that we should have a mutex in Journal to serialize access to j.cjournal?

File: journal/read.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	// TODO: the timeout logic here isn't very good.
+	pollDone := make(chan bool, 1)
+	go func() {
+	poll:
+		for {
+			select {
+			case <-pollDone:
+				break poll
+			default:
+				events <- r.journal.Wait(uint64(time.Duration(2) * time.Second))
```
Comment (vcaputo) [inline journal/read.go]: This doesn't seem safe to me, unless I'm missing something we can't safely enter the sd_journal C api from multiple goroutines on the same journal instance.  This is basically a continuation of my concern mentioned above about a lack of mutex serializing concurrent access to r.journal.cjournal.

File: journal/journal.go (hunk 1 above)
Comment (philips) [inline journal/journal.go]: ack, this seems correct.

Comment (vcaputo) [inline journal/journal.go]: I put this together to address the misuse of the C api:
https://github.com/vcaputo/go-systemd/commit/a6aa1c6141da049a42704a7eabf92b502032fd22

/cc @jonboulle @philips

Comment (jonboulle) [inline journal/journal.go]: @vcaputo I am swamped with other things, do you want to just put that up as a PR and we can close mine?

Comment (vcaputo) [inline journal/journal.go]: Sure, we can do that.

Comment (vcaputo) [inline journal/journal.go]: https://github.com/coreos/go-systemd/pull/110

still in rfc form of course

**Comment by jonboulle** [comment]
#110

**Pull request closed without merging**